mass_editing/migrations/14.0.1.0.0/pre-migrate.py

```diff
@@ -0,0 +1,83 @@
+# Copyright (C) 2020 - Iván Todorovich <ivan.todorovich@gmail.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openupgradelib import openupgrade
+
+
+@openupgrade.migrate()
+def migrate(env, version):
+    if openupgrade.table_exists(env.cr, "mass_editing"):
+        migrate_mass_editing(env)
+
+
+def migrate_mass_editing(env):
+    """ Migrates mass.editing to ir.actions.server """
+    # Remove FK (mass_editing_id)
+    openupgrade.remove_tables_fks(env.cr, ["mass_editing_line"])
+    # Create ir.actions.server for each mass.editing
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE ir_act_server
+        ADD COLUMN __legacy_mass_editing_id__       int4,
+        ADD COLUMN __legacy_ref_ir_act_window_id__  int4
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        INSERT INTO ir_act_server (
+            __legacy_mass_editing_id__,
+            __legacy_ref_ir_act_window_id__,
+            name,
+            type,
+            usage,
+            binding_type,
+            activity_user_type,
+            state,
+            model_id,
+            model_name,
+            domain
+        )
+        SELECT
+            me.id,
+            me.ref_ir_act_window_id,
+            COALESCE(me.action_name, me.name),
+            'ir.actions.server',
+            'ir_actions_server',
+            'action',
+            'specific',
+            'mass_edit',
+            me.model_id,
+            mo.model,
+            me.domain
+        FROM mass_editing me
+        LEFT JOIN ir_model mo ON (me.model_id = mo.id)
+        """,
+    )
+    # Delete fields that no longer exist (ondelete didn't exist before)
+    openupgrade.logged_query(
+        env.cr,
+        """
+        DELETE FROM mass_editing_line
+        WHERE field_id IS NULL
```

Comment:
Don't delete anything as explained before

Reply:
If I don't do this I get this error:

```
odoo.sql_db: bad query: ALTER TABLE "mass_editing_line" ADD FOREIGN KEY ("field_id") REFERENCES "ir_model_fields"("id") ON DELETE cascade
ERROR: insert or update on table "mass_editing_line" violates foreign key constraint "mass_editing_line_field_id_fkey"
DETAIL:  Key (field_id)=(15266) is not present in table "ir_model_fields".
```

I think, as some fields have been removed from previous odoo version (normally), some mass editing lines are left in this inconsistent state. But if the field have been removed.. I think it's ok to remove the mass.editing.line, don't you think?

Reply:
The error you are mentioning doesn't relates to this code. Removing null lines is not the same as having a FK constraint. What you mention can happen, but maybe the solution is to keep `ondelete="cascade"` on the corresponding field.

Reply:
I'm not sure I understand. `ondelete="cascade"` didn't exist before, I introduced it here.

After migrating the database, some `mass.editing.line` records are pointing to non-existing fields (because they have been removed by the corresponding modules)

Honestly I don't know why the missing field id (15266) is still there, the default ondelete value being `ondelete="set null"`, it should've set it to null..  however this is what we've received from upgrade.odoo.com

In any case you get the point. After the migration you're left with mass.editing.lines that will either:
- point to missing field ids (but let's assume that's out of scope, and consider this a bug in upgrade.odoo.com)
- field_id will be NULL

What's the point in keeping these lines? They fields don't exist anymore and they will only prevent the "field required" database constraint from being created.

I'll remove it if you like, I just think others might have the same issue without these lines

Reply:
OK, I get your point as well and it's OK to keep such code. OpenUpgrade doesn't remove such field entries.